loki/frontend/fparser.py

```diff
@@ -138,9 +139,14 @@ def parse_fparser_expression(source, scope):
     # Wrap source in brackets to make sure it appears like a valid expression
     # for fparser, and strip that Parenthesis node from the ast immediately after
     ast = Fortran2003.Primary('(' + source + ')').children[1]
-    _ir = parse_fparser_ast(ast, source, scope=scope)
+
+    # We parse the standalone expression with a dummy scope, to
+    # avoid overriding existing type info from the given scope.
+    _ir = parse_fparser_ast(ast, source, scope=Scope())
+
     # TODO: use rescope visitor for this
-    rescope_map = {v: v.clone(scope=scope) for v in FindTypedSymbols().visit(_ir)}
+    # In the rescoping we force using the existing type info via `type=None`
+    rescope_map = {v: v.clone(scope=scope, type=None) for v in FindTypedSymbols().visit(_ir)}
     _ir = SubstituteExpressions(rescope_map).visit(_ir)
```

Comment:
Might make sense to do this by using corresponding functionality that's already available (and thus take care of the TODO in the process), e.g.:
```
{v: v.rescope(scope=scope)}
```
or, even better because the parse expression may actually be using symbols from a parent scope:
```
_ir = AttachScopes().visit(_ir, scope=scope)
```

Reply:
Agreed, the second suggestion looks much neater and works very well. Updated.